src/track/beats.cpp

```diff
@@ -53,3 +190,666 @@ double Beats::findNBeatsFromSample(double fromSample, double beats) const {
+            std::lower_bound(m_beats.cbegin(), m_beats.cend(), startBeat, BeatLessThan);
+
+    BeatList::const_iterator lastBeat =
+            std::upper_bound(m_beats.cbegin(), m_beats.cend(), stopBeat, BeatLessThan);
+
+    if (curBeat >= lastBeat) {
+        return std::unique_ptr<BeatIterator>();
+    }
+    return std::make_unique<BeatIterator>(curBeat, lastBeat);
+}
+
+bool Beats::hasBeatInRangeNew(double startSample, double stopSample) const {
+    QMutexLocker locker(&m_mutex);
+    if (!isValid() || startSample > stopSample) {
+        return false;
+    }
+    double curBeat = findNextBeat(startSample);
+    if (curBeat <= stopSample) {
+        return true;
+    }
+    return false;
+}
+
+double Beats::getBpmNew() const {
+    QMutexLocker locker(&m_mutex);
+    if (!isValid())
+        return -1;
+    return m_dCachedBpm;
+}
+double Beats::getBpmRangeNew(FrameNum startFrame, FrameNum stopFrame) const {
+    QMutexLocker locker(&m_mutex);
+    if (!isValid())
+        return -1;
+    track::io::Beat startBeat, stopBeat;
+    startBeat.set_frame_position(startFrame);
+    stopBeat.set_frame_position(stopFrame);
+    return calculateBpm(startBeat, stopBeat);
+}
+
+double Beats::getBpmAroundPositionNew(FrameNum curFrame, int n) const {
+    QMutexLocker locker(&m_mutex);
+    if (!isValid())
+        return -1;
+
+    // To make sure we are always counting n beats, iterate backward to the
+    // lower bound, then iterate forward from there to the upper bound.
+    // a value of -1 indicates we went off the map -- count from the beginning.
+    double lower_bound = findNthBeatNew(curFrame, -n);
+    if (lower_bound == -1) {
+        lower_bound = m_beats.first().frame_position();
+    }
+
+    // If we hit the end of the beat map, recalculate the lower bound.
+    //double upper_bound = findNthBeat(lower_bound, n * 2);
+    double upper_bound = findNthBeatNew(lower_bound, n);
+    if (upper_bound == -1) {
+        upper_bound = m_beats.last().frame_position();
+        //lower_bound = findNthBeat(upper_bound, n * -2);
+        lower_bound = findNthBeatNew(upper_bound, n * -1);
+        // Super edge-case -- the track doesn't have n beats!  Do the best
+        // we can.
+        if (lower_bound == -1) {
+            lower_bound = m_beats.first().frame_position();
+        }
+    }
+
+    track::io::Beat startBeat, stopBeat;
+    startBeat.set_frame_position(lower_bound);
+    stopBeat.set_frame_position(upper_bound);
+    return calculateBpm(startBeat, stopBeat);
+}
+
+void Beats::addBeatNew(FrameNum beatFrame) {
+    QMutexLocker locker(&m_mutex);
+    track::io::Beat beat;
+    beat.set_frame_position(beatFrame);
+    BeatList::iterator it = std::lower_bound(
+            m_beats.begin(), m_beats.end(), beat, BeatLessThan);
+
+    // Don't insert a duplicate beat. TODO(XXX) determine what epsilon to
+    // consider a beat identical to another.
+    if (it->frame_position() == beat.frame_position())
+        return;
```

Comment:
Never omit brackets. It is an easy way for bugs to slip through.
```suggestion
    if (it->frame_position() == beat.frame_position()) {
        return;
    }
```

Reply:
Done. Is not possible to have clang informing about them?